### PR TITLE
Add proto3 specific deserialization that requires fewer SRTP constraints

### PR DESCRIPTION
**Pull request by 7sharp9**

### CODE REVIEW

**Comment by jhugard** [comment]
Going this route, the proto2 specific code should probably go in a Proto2 module.  Breaking change, yes, but cleaner.

Also, would it be possible to define common functions in terms of the other module to eliminate duplication?  E.g., in module Proto2, define `let inline fromZeroCopyBuffer m zcb = Proto3.fromZeroCopyBuffer m zcb`?  As in this example, I'd make Proto3 the base case, and define Proto2 in terms of Proto3.

**Comment by 7sharp9** [comment]
I did think about sharing the core stuff but most of the functions share just a shard:

```
    module Shared =
        /// Deserialize a message from a ZeroCopyBuffer, given a default message.
        let inline fromZeroCopyBuffer f m zcb =
            zcb
            |> Utility.decodeBuffer
            |> f m
            |> Helpers.Proto2.decodeFixup
            
        /// Deserialize a length-delimited message from a ZeroCopyBuffer, given a default message.
        let inline fromZeroCopyBufferLengthDelimited f m zcb =
            zcb
            |> Utility.unpackLengthDelimited
            |> f m
            |> Helpers.Proto2.decodeFixup
            
        /// Deserialize a message from a length-delimited RawField, given a default message.
        let inline fromRawField f m (rawField:RawField) =
            let buf = 
                match rawField with
                | LengthDelimited (fieldId, buf) ->
                    buf
                | _ ->
                    raise <| SerializerException(sprintf "Expected LengthDelimited field, found %s" (rawField.GetType().Name) )
            buf
            |> ZeroCopyBuffer
            |> fromZeroCopyBuffer f m

    module Proto2 =
        /// Deserialize a message from a ZeroCopyBuffer, given a default message.
        let inline fromZeroCopyBuffer m zcb =
            Shared.fromZeroCopyBuffer Helpers.Proto2.deserializeFields m zcb
    
        /// Deserialize a message from a ZeroCopyBuffer, given a default message.
        /// Shorthand for Deserialize.fromZeroCopyBuffer.
        let inline fromZcb m zcb = fromZeroCopyBuffer m zcb
    
        /// Deserialize a length-delimited message from a ZeroCopyBuffer, given a default message.
        let inline fromZeroCopyBufferLengthDelimited m zcb =
            Shared.fromZeroCopyBufferLengthDelimited Helpers.Proto2.deserializeFields m zcb
           
        /// Deserialize a length-delimited message from a ZeroCopyBuffer, given a default message.
        /// Shorthand for Deserialize.fromZeroCopyBufferLengthDelimited.
        let inline fromZcbLD m zcb = fromZeroCopyBufferLengthDelimited m zcb
    
        /// Deserialize a message from a length-delimited RawField, given a default message.
        let inline fromRawField m (rawField:RawField) =
            Shared.fromRawField Helpers.Proto2.deserializeFields m rawField
    
        /// Deserialize a message from a length-delimited RawField, given a default message,
        /// and return Some(message).  Used to simplify the call-site when deserializing
        /// inner messages.
        let inline optionalMessage m rawField =
            Some (fromRawField m rawField)
    
        /// Deserialize a message from an ArraySegment, given a default message.
        let inline fromArraySegment m (buf:ArraySegment<byte>) =
            buf
            |> ZeroCopyBuffer
            |> fromZcb m
    
        /// Deserialize a length-delimited message from an ArraySegment, given a default message.
        let inline fromArraySegmentLengthDelimited m (buf:ArraySegment<byte>) =
            buf
            |> ZeroCopyBuffer
            |> fromZcbLD m
    
        /// Deserialize a length-delimited message from an ArraySegment, given a default message.
        /// Shorthand for Deserialize.fromArraySegmentLengthDelimited.
        let inline fromArraySegmentLD m buf = fromArraySegmentLengthDelimited m buf
    
        /// Deserialize a message from a byte array, given a default message.
        let inline fromArray m buf =
            buf
            |> ArraySegment
            |> fromArraySegment m
    
        /// Deserialize a length-delimited message from a byte array, given a default message.
        let inline fromArrayLengthDelimited m buf =
            buf
            |> ArraySegment
            |> fromArraySegmentLD m
    
        /// Deserialize a length-delimited message from a byte array, given a default message.
        /// Shorthand for Deserialize.fromArrayLengthDelimited.
        let inline fromArrayLD m buf = fromArrayLengthDelimited m buf
```
Code like that starts to have quite bad readability for the code sharing it offers

**Comment by 7sharp9** [comment]
Ive updated the PR to have as much sharing as possible.  I have also placed proto2/3 in separate modules.  

Deserialization now has to be explicit.  As the scope of the deserialization is code generation this is probably a good thing in light of possible proto2/3 confusion, all it takes is opening the required module to access what you need.

**Comment by jhugard** [comment]
Other than the above comments, the changes look very clean.  Thank you!

**Comment by jhugard** [comment]
> I have nothing against the optional support but having the fields in the record definition seems very hacky thats all.

I couldn't figure a way to have those fields in each record, but not need to provide them when instantiating the record.  Did you have an idea here that would be cleaner?

**Comment by 7sharp9** [comment]
The simplest thing is to check the decoderRing and use field 0 like proto2 except have the behavior of not throwing is there no decoder.  Then its up to the code gen to add the zero field for unknowns which should be trivial.  the generator could be given parameters whether to create an unknown field section in the record definition.

**Comment by 7sharp9** [comment]
Fundamentally record construction makes the unknowns tricky, the only thing I could think of was generating a static construction method with the last param being optional.

**Comment by 7sharp9** [comment]
With a class thats generated it would be easy to just have a `ResizeArray` etc and a method to get unknown fields, but theres no mechanism available with a record.

Another area that I thought could be unperformant was any proto messages that has a large quantity or repeated fields.

**Comment by 7sharp9** [comment]
@jhugard I added a few tests as requested.

File: Serialization.Test/TestSerialization.fs [hunk 1]
```diff
@@ -175,4 +175,87 @@ module RecordSerialization =
...
+                    0, fun m rawField -> { m with _unknownFields = rawField :: m._unknownFields } : Proto3Message
+                    1, fun m rawField -> { m with id = rawField |> Decode.toInt32 } : _
+                    2, fun m rawField -> { m with name = rawField |> Decode.toString } : _
+                ]
+                |> Map.ofList
+    
+            static member DecodeFixup m =
+                { m with _unknownFields = List.rev m._unknownFields }
+
+            static member RequiredFields =
```
Comment (jhugard) [inline Serialization.Test/TestSerialization.fs]: Superfluous.  RequiredFields should be removed for proto3.

File: Serialization.Test/TestSerialization.fs [hunk 2]
```diff
@@ -175,4 +175,87 @@ module RecordSerialization =
...
+    type Proto3Message  = {
+        id : int32
+        name : string
+        _unknownFields : RawField list
+        }
+        with
+            static member Default = {
+                id = 0
+                name = ""
+                _unknownFields = List.empty
```
Comment (jhugard) [inline Serialization.Test/TestSerialization.fs]: Should you remove this to show usage without unknown fields?  Or, comment this is optional?

File: Serialization.Test/TestSerialization.fs [hunk 3]
```diff
@@ -175,4 +175,87 @@ module RecordSerialization =
...
+                ]
+                |> Map.ofList
+    
+            static member DecodeFixup m =
+                { m with _unknownFields = List.rev m._unknownFields }
+
+            static member RequiredFields =
+                [ 1; 2 ] |> Set.ofList
+
+            static member UnknownFields m =
```
Comment (jhugard) [inline Serialization.Test/TestSerialization.fs]: Same as above: should you remove this to show usage without unknown fields?  Or, comment this is optional?

**Comment by jhugard** [comment]
Better, but suggest a touch of cleanup to show your desired Proto3 usage.

File: Serialization.Test/TestSerialization.fs (hunk 2 above)
Comment (7sharp9) [reply Serialization.Test/TestSerialization.fs]: I'll add documentation as this is up the the code generator to enforce really.  There are a few parts where you can produce proto code that will fail at runtime like having no ring 0 for unknown fields, or forgetting to reverse fields in the fixup, so I think some docs will do for now.  A further cleanup in a future PR could tighten things up a little.

**Comment by 7sharp9** [comment]
OK, I have now added some details in the proto3 test types.

**Comment by jhugard** [comment]
Great work. Thank you!

**Pull request merged**